registry/registryclient/client.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+
+package registryclient
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/opentofu/libregistry/registry/providerregistryprotocol"
+	"github.com/opentofu/libregistry/registry/servicediscovery"
+)
+
+type ClientOpt func(cfg *config) error
+
+func NewClient(opts ...ClientOpt) (Client, error) {
+	cfg := config{}
+	for _, opt := range opts {
+		if err := opt(&cfg); err != nil {
+			return nil, err
+		}
+	}
+	cfg.applyDefaults()
+	return &client{
+		cfg,
+	}, nil
+}
+
+func WithHTTPClient(client *http.Client) ClientOpt {
+	return func(cfg *config) error {
+		cfg.client = client
+		return nil
+	}
+}
+
+func WithEndpoint(endpoint string) ClientOpt {
+	return func(cfg *config) error {
+		cfg.endpoint = endpoint
+		return nil
+	}
+}
+
+type config struct {
+	endpoint string
+	client   *http.Client
+}
+
+func (c *config) applyDefaults() {
+	if c.client == nil {
+		c.client = http.DefaultClient
+	}
+	if c.endpoint == "" {
+		c.endpoint = "https://registry.opentofu.org"
+	}
+}
+
+type client struct {
+	cfg config
+}
+
+func (c client) ServiceDiscovery(ctx context.Context) (DiscoveredClient, error) {
+	sd, err := servicediscovery.NewClient(
+		servicediscovery.WithEndpoint(c.cfg.endpoint),
+		servicediscovery.WithHTTPClient(c.cfg.client),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize service discovery client (%w)", err)
+	}
+	resp, err := sd.ServiceDiscovery(ctx, servicediscovery.Request{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to perform service discovery (%w)", err)
+	}
+
+	if resp.ProvidersV1 == "" {
+		return nil, fmt.Errorf("no providers endpoint found")
```

Comment:
The package name `registryclient` suggested to me that this particular component was intended to be broader than just _provider_ registries, and would potentially grow to support module registries in future too. If that's true, then I would've expected this to support a registry that only supports one of the two current registry protocols, such as only offering a module registry and not also a provider registry.

If your intention was for this to be only for provider registry then perhaps the package name could be more specific to allow for adding a separate package for module registry client later?

Alternatively, I imagine a potential design where this `ServiceDiscovery` method fails only if the overall service discovery protocol failed in itself, regardless of which specific services were returned, and then the returned `DiscoveredClient` object is a struct with a method like `ProviderClient() (providerregistryprotocol.Client, error)` where _this_ `error` is where we'd report "no providers endpoint found", since the caller has then explicitly requested that particular service.

All of this is loose ideas, not hard objection. Seems like there's a few different approaches here with some different tradeoffs, so I'm not meaning to try to force any particular design.